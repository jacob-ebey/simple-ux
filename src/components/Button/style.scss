@import '../variables';

.button {
  display: inline-block;
  border: $border-width solid $color-dark;
  padding: 5px;
  background-color: $color-light;
  color: $color-dark;
  text-decoration: none;
  cursor: pointer;
  user-select: none;

  &-disabled {
    color: $color-gray;
  }

  &:hover, &:focus {
    border: $border-width solid $color-dark;
    background-color: $color-light-darker;
  }

  &:active {
    border-color: $color-lessdark;
    background-color: $color-light-darkest;
  }

  &-primary {
    border: $border-width solid $color-primary;

    &:hover, &:focus {
      border-color: $color-primary-dark;
    }
  }

  &-secondary {
    border: $border-width solid $color-secondary;

    &:hover, &:focus {
      border-color: $color-secondary-dark;
    }
  }

  &-raise, &-raise-primary, &-raise-secondary {
    &:hover, &:focus {
      box-shadow: 0 0.5em 0.5em -0.5em $color-dark;
      transform: translateY(-0.25em);
    }
  }

  &-pulse {
    &:hover, &:focus {
      animation: -pulse 0.5s;
      box-shadow: 0 0 0 0.5em rgba($color-light, 0);
    }
    @keyframes -pulse {
      0% { box-shadow: 0 0 0 0 $color-dark; }
    }
  }

  &-pulse-primary {
    &:hover, &:focus {
      animation: -pulse-primary 0.5s;
      box-shadow: 0 0 0 0.5em rgba($color-light, 0);
    }
    @keyframes -pulse-primary {
      0% { box-shadow: 0 0 0 0 $color-primary; }
    }
  }

  &-pulse-secondary {
    &:hover, &:focus {
      animation: -pulse-secondary 0.5s;
      box-shadow: 0 0 0 0.5em rgba($color-light, 0);
    }
    @keyframes -pulse-secondary {
      0% { box-shadow: 0 0 0 0 $color-secondary; }
    }
  }
}