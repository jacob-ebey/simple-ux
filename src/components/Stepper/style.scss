@import '../variables';

.stepper {
  border: $border-width solid $color-dark;
  padding: $content-padding / 2;

  &-title {
    font-size: 1.5em;
  }

  &-body {
    border: $border-width solid $color-dark;
    margin: $content-padding / 2 0;
  }

  &-footer {
    .button {
      margin-right: $content-padding / 2;
    }
  }
}