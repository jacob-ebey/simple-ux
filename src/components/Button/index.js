import { h, Component } from 'preact' // eslint-disable-line no-unused-vars

import { merge, merge2, mod, raise } from '../tools'

import './style.scss'

/**
 * Button Component
 */
export default class Button extends Component {
  render (props) {
    const { animation, color, disabled, onClick, class: className, children, ...rest } = props

    return (
      <a
        {...rest}
        class={
          merge(
            'button',
            mod('button', color),
            mod('button', disabled && 'disabled'),
            mod('button', merge2(animation, color)),
            className
          )
        }
        onClick={onClick && this._handleClick}
      >
        {children}
      </a>
    )
  }

  _handleClick = (ev) => {
    const { onClick, disabled } = this.props

    raise(onClick && !disabled, () => {
      ev.preventDefault()

      onClick(ev)
    })
  }
}
