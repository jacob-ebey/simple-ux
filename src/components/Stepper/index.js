import { h, Component } from 'preact' // eslint-disable-line no-unused-vars

import Button from '../Button'
import { merge, childArray } from '../tools'

import './style.scss'

/**
 * Stepper Component
 */
export default class Stepper extends Component {
  state = {
    step: 0
  }

  get canIncrement () {
    const { step } = this.state
    const { children } = this.props

    const count = childArray(children).length

    return step < count - 1
  }

  get canDecrement () {
    const { step } = this.state

    return step > 0
  }

  render (props) {
    const { nextLabel, previousLabel, class: className, children, ...rest } = props

    return (
      <div class={merge('stepper', className)} {...rest}>
        <div class='stepper-body'>
          {this._renderStep()}
        </div>
        <div class='stepper-footer'>
          <Button data-test='decrement' disabled={this.canDecrement} onClick={this._decrementStep}>
            {previousLabel}
          </Button>
          <Button data-test='increment' disabled={this.canIncrement} onClick={this._incrementStep}>
            {nextLabel}
          </Button>
        </div>
      </div>
    )
  }

  _renderStep = () => {
    const { step } = this.state
    const { children } = this.props

    const c = childArray(children)

    return c[step]
  }

  _incrementStep = () => {
    const { step } = this.state

    if (this.canIncrement) {
      this.setState({ step: step + 1 })
    }
  }

  _decrementStep = () => {
    const { step } = this.state

    if (this.canDecrement) {
      this.setState({ step: step - 1 })
    }
  }
}

Stepper.defaultProps = {
  nextLabel: 'Next',
  previousLabel: 'Previous'
}
