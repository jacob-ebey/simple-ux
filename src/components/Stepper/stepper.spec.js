/* eslint-env jest */
import { h } from 'preact' // eslint-disable-line no-unused-vars
import { shallow } from 'preact-render-spy'

import Button from '../Button'

import Stepper from './'

describe('Stepper', () => {
  describe('- snapshots -', () => {
    it('can render', () => {
      const wrapper = shallow(<Stepper />)
      expect(wrapper).toMatchSnapshot()
    })

    it('can add custom class', () => {
      const wrapper = shallow(<Stepper class='custom' />)
      expect(wrapper).toMatchSnapshot()
    })

    it('only renders first child', () => {
      const wrapper = shallow(
        <Stepper>
          <span>Child 1</span>
          <span>Child 2</span>
        </Stepper>
      )
      expect(wrapper).toMatchSnapshot()
    })

    it('only renders second child', () => {
      const wrapper = shallow(
        <Stepper>
          <span>Child 1</span>
          <span>Child 2</span>
        </Stepper>
      )

      wrapper.setState({ step: 1 })
      expect(wrapper).toMatchSnapshot()
    })

    it('can render title', () => {
      const wrapper = shallow(<Stepper title='ROFL' />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('- logic -', () => {
    it('can increment', () => {
      const wrapper = shallow(
        <Stepper>
          <span>Child 1</span>
          <span>Child 2</span>
        </Stepper>
      )

      wrapper.find(<Button data-test='increment' />).simulate('click')
      expect(wrapper.state('step')).toBe(1)
    })

    it('does not increment', () => {
      const wrapper = shallow(
        <Stepper>
          <span>Child 1</span>
        </Stepper>
      )
      const increment = wrapper.find(<Button data-test='increment' />)
      increment.simulate('click')
      expect(wrapper.state('step')).toBe(0)
    })

    it('does not increment when custom', () => {
      const wrapper = shallow(
        <Stepper canIncrement={() => false}>
          <span>Child 1</span>
          <span>Child 2</span>
        </Stepper>
      )
      const increment = wrapper.find(<Button data-test='increment' />)
      increment.simulate('click')
      expect(wrapper.state('step')).toBe(0)
    })

    it('can decrement', () => {
      const wrapper = shallow(
        <Stepper>
          <span>Child 1</span>
          <span>Child 2</span>
        </Stepper>
      )
      wrapper.setState({ step: 1 })
      const decrement = wrapper.find(<Button data-test='decrement' />)
      decrement.simulate('click')
      expect(wrapper.state('step')).toBe(0)
    })

    it('does not decrement', () => {
      const wrapper = shallow(
        <Stepper>
          <span>Child 1</span>
        </Stepper>
      )
      const decrement = wrapper.find(<Button data-test='decrement' />)
      decrement.simulate('click')
      expect(wrapper.state('step')).toBe(0)
    })

    it('does not decrement when custom', () => {
      const wrapper = shallow(
        <Stepper canDecrement={() => false}>
          <span>Child 1</span>
          <span>Child 2</span>
        </Stepper>
      )
      wrapper.setState({ step: 1 })
      const decrement = wrapper.find(<Button data-test='decrement' />)
      decrement.simulate('click')
      expect(wrapper.state('step')).toBe(1)
    })
  })
})
