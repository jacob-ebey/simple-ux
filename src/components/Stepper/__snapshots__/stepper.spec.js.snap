// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper - snapshots - can add custom class 1`] = `
preact-render-spy (1 nodes)
-------
<div class="stepper custom">
  <div class="stepper-body"></div>
  <div class="stepper-footer">
    <Button
      data-test="decrement"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Previous
    </Button>
    <Button
      data-test="increment"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Next
    </Button>
  </div>
</div>

`;

exports[`Stepper - snapshots - can render 1`] = `
preact-render-spy (1 nodes)
-------
<div class="stepper">
  <div class="stepper-body"></div>
  <div class="stepper-footer">
    <Button
      data-test="decrement"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Previous
    </Button>
    <Button
      data-test="increment"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Next
    </Button>
  </div>
</div>

`;

exports[`Stepper - snapshots - can render title 1`] = `
preact-render-spy (1 nodes)
-------
<div class="stepper">
  <div class="stepper-title">ROFL</div>
  <div class="stepper-body"></div>
  <div class="stepper-footer">
    <Button
      data-test="decrement"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Previous
    </Button>
    <Button
      data-test="increment"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Next
    </Button>
  </div>
</div>

`;

exports[`Stepper - snapshots - only renders first child 1`] = `
preact-render-spy (1 nodes)
-------
<div class="stepper">
  <div class="stepper-body">
    <span>Child 1</span>
  </div>
  <div class="stepper-footer">
    <Button
      data-test="decrement"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Previous
    </Button>
    <Button
      data-test="increment"
      disabled={false}
      onClick={[Function anonymous]}
    >
      Next
    </Button>
  </div>
</div>

`;

exports[`Stepper - snapshots - only renders second child 1`] = `
preact-render-spy (1 nodes)
-------
<div class="stepper">
  <div class="stepper-body">
    <span>Child 2</span>
  </div>
  <div class="stepper-footer">
    <Button
      data-test="decrement"
      disabled={false}
      onClick={[Function anonymous]}
    >
      Previous
    </Button>
    <Button
      data-test="increment"
      disabled={true}
      onClick={[Function anonymous]}
    >
      Next
    </Button>
  </div>
</div>

`;
